File: vol1/ch3/LeetCode155-3.cpp
typedef long long LL;
class MinStack {
    stack<LL> st;
    LL minval=-1;
public:
    MinStack() {  }

    void push(LL val) {
        if(!st.empty()) {       //st·Ç¿Õ
            LL d=val-minval;
            st.push(d);
            if(d<0) minval=val;
        }
        else {                  //stÎª¿Õ
            st.push(0);
            minval=val;
        }
    }

    void pop() {
        LL d=st.top(); st.pop();
        if(d<0) minval=minval-d;
    }
    
    int top() {
        if(st.top()<0) return minval;
        else return st.top()+minval;
    }
    
    int getMin() {
        if(st.empty()) return -1;
        else return minval;
    }
};

File: vol1/ch3/LeetCode155-3.py
class MinStack:
    def __init__(self):
        self.st=[]
        self.minval=-1

    def push(self, val: int) -> None:
        if len(self.st)>0:              #st非空
            d=val-self.minval
            self.st.append(d)
            if d<0: self.minval=val
        else:                           #st为空
            self.st.append(0)
            self.minval=val

    def pop(self) -> None:
        d=self.st.pop()
        if d<0:
            self.minval=self.minval-d

    def top(self) -> int:
        if self.st[-1]<0:return self.minval
        else:return self.st[-1]+self.minval

    def getMin(self) -> int:
        if len(self.st)==0: return -1
        else: return self.minval
